statefulMapConcat examples (#28611)
akka/akka: akka-docs/src/main/paradox/stream/operators/Source-or-Flow/statefulMapConcat.md

## Examples

In this first sample we keep a counter, and combine each element with an id that is unique for the stream materialization
(replicating the @ref:[zipWithIndex](zipWithIndex.md) operator):

Scala
:  @@snip [StatefulMapConcat.scala](/akka-docs/src/test/scala/docs/stream/operators/flow/StatefulMapConcat.scala) { #zip-with-index }

Java
:   @@snip [StatefulMapConcat.java](/akka-docs/src/test/java/jdocs/stream/operators/flow/StatefulMapConcat.java) { #zip-with-index }

In this sample we let the value of the elements have an effect on the following elements, if an element starts
with `blacklist:word` we add it to a black list and filter out any subsequent entries of `word`:

Scala
:  @@snip [StatefulMapConcat.scala](/akka-docs/src/test/scala/docs/stream/operators/flow/StatefulMapConcat.scala) { #blacklist }

Java
:   @@snip [StatefulMapConcat.java](/akka-docs/src/test/java/jdocs/stream/operators/flow/StatefulMapConcat.java) { #blacklist }

For cases where there is a need to emit elements based on the state when the stream ends, it is possible to add an extra

> element signalling the end of the stream before the `statefulMapConcat` operator.

Reviewer: Neat trick!

In this sample we collect all elements starting with the letter `b` and emit those once we have reached the end of the stream using

> as special end element. Here the end element is a special string to keep sample concise, in a real application it can make sense to rather use types.

Reviewer: ```suggestion
a special end element. The end element is a special string to keep the sample concise, in a real application it may make sense to use types instead.
```
